### DISCUSSION TITLE
Prefix n' Shorten not working because of uri slash

### DISCUSSION THREAD

**Discussion by flashlab** [YOURLS/YOURLS#3655] [Q&A] [answered]
If I enter `https://sho.rt/subpath/https://longurl.com/something.html`, I always been redirect to YOURLS_SITE(`https://sho.rt/subpath/`). What I want is to use prefix to quickly shorten urls.
So I did some debugging. In `yourls_get_request`, the `$_SERVER[ 'REQUEST_URI' ]` returned `/subpath/https:/longurl.com/something.html` which `/subpath/https://longurl.com/something.html` is expected.
The problem is `https://` turn into `https:/` with one slash missing.
I thought the problem maybe come from apache rewrite rules, here is my .htaccess file in YOURLS root folder:
```
# BEGIN YOURLS
<IfModule mod_rewrite.c>
RewriteEngine On
RewriteBase /u/
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule ^.*$ /u/yourls-loader.php [L]
</IfModule>
# END YOURLS
```

Can anyone help me to solve it?

=====
```
$request = preg_replace('/^(https?):\/(?!\/)/', "$1://", $request, 1);
```
This is the ugly solution.

**Comment by dgw** [comment]
I have YOURLS installed at the domain root, but I added a temporary copy of the files at `/subdir/` to test this. The only changes from my root `.htaccess` were adding `subdir/` to the `RewriteBase` and `RewriteRule` lines:

```apache
# BEGIN YOURLS
<IfModule mod_rewrite.c>
RewriteEngine On
RewriteBase /subdir/  # in /, this is `RewriteBase /`
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule ^.*$ /subdir/yourls-loader.php [L]  # in /, this is `RewriteRule ^.*$ /yourls-loader.php [L]
</IfModule>
# END YOURLS
```

And I changed `YOURLS_SITE` to include `subdir/` as well in `user/config.php`.

On both my root install and the temporary subdir copy, prefix-n-shorten worked as expected. Both `https://domain.tld/https://some-other.site` and `https://domain.tld/subdir/https://some-other.site` created a new shortlink and redirected as expected.

A quick search tells me that [`MergeSlashes`](https://httpd.apache.org/docs/2.4/mod/core.html#mergeslashes) is probably enabled in your Apache configuration. This feature (I guess they consider it a feature) was added in Apache 2.4.39, and it's on by default.

My site (on shared hosting) is a bit behind that version; I'll have to keep `MergeSlashes` in mind for my own installation whenever the host upgrades things. It doesn't look like I can turn it off at the `.htaccess` level, so I just might have to ask support to change the vhost configuration for me if the eventual Apache update breaks this feature… Thanks for putting it on my radar. 🙂

----

_An alternative cause could be `ProxyPass` canonicalizing proxied URLs, if you are using `mod_proxy` in front of YOURLS. [The `nocanon` keyword will suppress this](https://httpd.apache.org/docs/2.4/mod/mod_proxy.html#proxypass), but the recently-added `MergeSlashes` directive above is the more likely cause._

**Comment by flashlab** [comment]
@dgw  I checked my apache info, here is the output:
```
➜  ~ /usr/local/apache/bin/apachectl -V
Server version: Apache/2.4.55 (Unix)
Server built:   Jun 29 2023 05:10:23
Server's Module Magic Number: 20120211:126
Server loaded:  APR 1.5.2, APR-UTIL 1.5.4, PCRE 8.37 2015-04-28
Compiled using: APR 1.5.2, APR-UTIL 1.5.4, PCRE 8.37 2015-04-28
Architecture:   64-bit
Server MPM:     worker
  threaded:     yes (fixed thread count)
    forked:     yes (variable process count)
Server compiled with....
 -D BIG_SECURITY_HOLE
 -D APR_HAS_SENDFILE
 -D APR_HAS_MMAP
 -D APR_HAVE_IPV6 (IPv4-mapped addresses enabled)
 -D APR_USE_SYSVSEM_SERIALIZE
 -D APR_USE_PTHREAD_SERIALIZE
 -D SINGLE_LISTEN_UNSERIALIZED_ACCEPT
 -D APR_HAS_OTHER_CHILD
 -D AP_HAVE_RELIABLE_PIPED_LOGS
 -D DYNAMIC_MODULE_LIMIT=256
 -D HTTPD_ROOT="/root/daily_build/510_10/5.1.0/NasMgmt/HTTP/apache_install_tmp"
 -D SUEXEC_BIN="/root/daily_build/510_10/5.1.0/NasMgmt/HTTP/apache_install_tmp/bin/suexec"
 -D DEFAULT_PIDLOG="logs/httpd.pid"
 -D DEFAULT_SCOREBOARD="logs/apache_runtime_status"
 -D DEFAULT_ERRORLOG="logs/error_log"
 -D AP_TYPES_CONFIG_FILE="/etc/config/apache/mime.types"
 -D SERVER_CONFIG_FILE="/etc/config/apache/apache.conf"
```
It seems `MergeSlashes` was not been shutdown.If so, I'll have to use the temporary solution, Thanks for your help.

**Reply by dgw** [reply, marked as answer]
FYI, if you were looking for information about it in the compilation options, `MergeSlashes` is not set at compile time. You are running a version newer than when the feature was added, so if you do not want it enabled, you'll have to add the option `MergeSlashes off` into one of the supported configuration files (server config or vhost config).

**Reply by flashlab** [reply]
@dgw I've figure out it was `ProxyPass` that do the trick in my situation.
After append `nocanon` within `ProxyPass`, everything works fine.
Just want you to know about it.
******
The sad story is: `nocanon` affect mediawiki api requests, so I have to use the ugly solution for now T.T
